urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/com/kokakiwi/mclauncher/core/updater/UpdaterWorker.java

```diff
@@ -0,0 +1,288 @@
+        File dest = null;
+        Type type = null;
+        
+        try
+        {
+            // Load libraries URLs
+            dest = new File(api.getMinecraftDirectory(), "bin");
+            type = Type.LIBRARY;
+            for (final String librariesFile : api.getConfig().getStringList(
+                    "updater.libraries"))
+            {
+                final URL url = new URL(librariesFile);
+                
+                final GameFile file = new GameFile(url, dest, type);
+                api.getUpdater().getGameFiles().add(file);
+            }
+            
+            // Load native URL
+            dest = new File(dest, "natives");
+            {
+                final URL url = new URL(api.getConfig().getString(
+                        "updater.natives." + SystemUtils.getSystemOS().name()));
+                type = Type.NATIVE;
+                final GameFile file = new GameFile(url, dest, type);
+                api.getUpdater().getGameFiles().add(file);
+            }
+            
+            // Load additionnals URLs
+            dest = new File(api.getMinecraftDirectory(), "bin");
+            type = Type.ADDITIONNAL;
+            for (final String additionnalFile : api.getConfig().getStringList(
+                    "updater.additionnals"))
+            {
+                final URL url = new URL(additionnalFile);
+                
+                final GameFile file = new GameFile(url, dest, type);
+                api.getUpdater().getGameFiles().add(file);
+            }
+        }
+        catch (final Exception e)
+        {
+            e.printStackTrace();
+        }
+    }
+    
+    public static class GameFile
+    {
+        private final URL         url;
+        private final File        dest;
+        private final Type        type;
+        
+        private HttpURLConnection connection = null;
+        
+        private float             speed      = 0;
+        private long              downloaded = 0;
+        private long              total      = 0;
+        
+        private int               percentage = 0;
+        
+        public GameFile(URL url, File dest, Type type) throws Exception
+        {
+            this.url = url;
+            this.dest = dest;
+            this.type = type;
+            
+            dest.mkdirs();
+        }
+        
+        public void init() throws Exception
+        {
+            connection = (HttpURLConnection) url.openConnection();
+            connection.setDefaultUseCaches(false);
+            
+            total = connection.getContentLength();
+        }
+        
+        public void download(LauncherAPI api) throws Exception
+        {
+            System.out.println("Downloading '" + getFileName() + "'...");
+            
+            long lastTime = System.currentTimeMillis();
+            
+            final InputStream in = connection.getInputStream();
+            final OutputStream out = new FileOutputStream(new File(dest,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *out*. The resource is referenced at line 162. The resource is closed at line 184. There are other execution paths that do not contain closure statements, for example, when InputStream.read throws an exception. Ensure *out* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).